### ISSUE TITLE
Packaging guidelines

### ISSUE THREAD

**Issue by untitaker** [untitaker/vdirsyncer#352]
I think it would make sense to document a few things for packaging vdirsyncer.

## How to test vdirsyncer

Many packagers try to run the full testsuite on a vdirsyncer installation, in order to find bugs with their setup. But: 99% of all packaging-related bugs are caught by seeing if `vdirsyncer --help` works correctly.

The testsuite is not suitable for packaging tests because:

* It uses [hypothesis](http://hypothesis.readthedocs.org/) to randomly generate example input data. Automatic test generation is a really powerful tool for finding encoding-related bugs, but it also means that tests may randomly fail.

* It launches CardDAV/CalDAV servers for integration testing. In other words, it's not just for finding bugs in vdirsyncer, but also elsewhere. Why would vdirsyncer-packagers care if a server introduced a bug that makes the tests fail?

* It requires very recent dependencies. Older pytest, hypothesis or Radicale versions may fail with bizzarre errors. This is especially a problem if packagers try to use their system-packages for installing test dependencies instead of just using a virtualenv and pip (I don't see a reason why that would be a problem TBH) See also #334 

This seems to be the sanest approach to testing vdirsyncer as a packager: https://github.com/Homebrew/homebrew/pull/41208 i.e. just create an example config and run a simple sync. Even a simple `vdirsyncer --help` catches 80% of errors IMO.

## Updating vdirsyncer

[The Hypothesis Packaging Guidelines](https://hypothesis.readthedocs.org/en/release/packaging.html) say that package maintainers who are unable to keep up with its release cycle are doing the community a disservice. While I think there are nicer ways to phrase that (especially towards people who are likewise doing a voluntary service), I am worried that vdirsyncer will find itself in a similarly problematic situation once [Debian starts to package it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=745027). Packages usually have been very quickly updated otherwise.

Apart from that we should probably provide an RSS feed for new releases. GitHub has a tags RSS feed that is mentioned [in the docs](https://vdirsyncer.readthedocs.org/en/latest/changelog.html), I guess we could just use that.

## Enforcing dependencies' minimum required versions

Vdirsyncer has a lower bound on the versions of some dependencies. As far as I understand those lower bounds have been completely ignored so far by downstream. Perhaps there's just no way to fix this situation. Again I'm more worried about Debian than any other distro here, but maybe it'll be fine.

--

Seeking feedback from

* @hobarrera @eolianoe (ArchLinux)
* @0-wiz-0 (pkgsrc)
* @mbaldessari @mathstuf (Fedora)
* @damiencassou @matthiasbeyer (nix)
* @benjaminfrank (Homebrew)
* @lfam (GNU + Guix)

**Comment by hobarrera** [comment]
> The Hypothesis Packaging Guidelines say that package maintainers who are unable to keep up with its release cycle are doing the community a disservice. While I think there are nicer ways to phrase that (especially towards people who are likewise doing a voluntary service),

It may sound harsh, but it does have a lot of truth. A package providing an older version, is exposing users to bugs that had been fixed and to a bad time. They also force you to deal with bug reports for older versions, etc.

Regarding Debian, if we find that an older verision keeps getting packaged, and this results in lots of users reporting invalid bugs, maybe it would be best to actually *recommend* in the documentation that users install via pip+virtualenv.

**Comment by lfam** [comment]
On Tue, Mar 01, 2016 at 01:45:39PM -0800, Markus Unterwaditzer wrote:
> I think it would make sense to document a few things for packaging
> vdirsyncer.

Thanks for asking us about this topic.

It seems that you are asking us to stop running your test suite. I
believe that letting other people run your test suite regularly will
help you find bugs more quickly and effectively.

You've given some reasons why packagers might have difficulty running
your test suite, but I'm still not sure why you'd want us to stop. Is it
creating extra work for you, or slowing down development in some way?
It's very common for software projects to distribute their source code
with a test suite.

I'd like to have a dialogue about this, and possibly come to a
compromise. In my opinion, the worst case outcome is that you stop
distributing tests in your release tarballs. I think that would have a
negative impact on the quality of vdirsyncer in the long run.

> ## How to test vdirsyncer
> 
> Many packagers try to run the full testsuite on a vdirsyncer
> installation, in order to find bugs with their setup. But: 99% of all
> packaging-related bugs are caught by seeing if `vdirsyncer --help`
> works correctly.

In my experience, '--help' usually exercises a very small part of a code
base. Is that not the case for vdirsyncer?

Running a (good and up-to-date) test suite does help catch problems with
our packaging, especially due to upstream portability bugs and version
mis-matches, but it also helps upstream by finding more serious bugs.

Here are 3 examples of recent bugs that we discovered (sometimes
independently of others) by running upstream test suites of *stable*
release tarballs. The first gnupg bug manifested on some implementations
of a specific architecture and not others. The ilmbase bug manifests on
32-bit Intel compatible only. In each case, the software compiled
without error! And to reiterate, in each case, upstream distributed this
code as a stable release.

gnupg: 14 of 35 tests failed: https://bugs.gnupg.org/gnupg/issue2229

gnupg: FAIL: gpgtar.test:
http://debbugs.gnu.org/cgi/bugreport.cgi?bug=22826

ilmbase: 32-bit floating point rounding problem (libreoffice depends on
ilmbase, so this matters to an OS distribution that wants to serve
desktop users):
https://lists.nongnu.org/archive/html/openexr-devel/2015-12/msg00001.html
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=815712

> The testsuite is not suitable for packaging tests because:
> 
> * It uses [hypothesis](http://hypothesis.readthedocs.org/) to randomly
> generate example input data. Automatic test generation is a really
> powerful tool for finding encoding-related bugs, but it also means
> that tests may randomly fail.

Can you clarify what would be bad about these tests failing when we
build these packages? Don't you want us to help you find those bugs?

If you mean that the tests are non-deterministic, I suggest that it's
not optimal to introduce non-determinism in the build process of your
software (I know you think the tests aren't part of that process, but
obviously I disagree). If you want to fuzz vdirsyncer, perhaps that
belongs in a separate body of tests. Or, you could add a note to INSTALL
instructing packagers to skip these tests.

Debian is already building >80% of their packages reproducibly [0], and
in Guix we are creating infrastructure to start treating
non-reproducible builds as a bug. I don't see this trend turning around;
it's too important for reliability and security.

What do you think? Have I misunderstood your concern about this point?

[0] https://tests.reproducible-builds.org/reproducible.html

> * It launches CardDAV/CalDAV servers for integration testing. In other
> words, it's not just for finding bugs in vdirsyncer, but also
> elsewhere. Why would vdirsyncer-packagers care if a server introduced
> a bug that makes the tests fail?

I see your point regarding this specific test. However, I don't think it
extends to the rest of the test suite.

It's not very different from bugs being introduced in another one of
vdirsyncer's build- or run-time dependencies. I want to know if an
update to e.g. 'click' is going to break our vdirsyncer package, or
expose a bug in vdirsyncer, etc. Nevertheless, as above, it could go in
a separate test suite or come with a caveat in INSTALL.

What do you think?

> * It requires very recent dependencies. Older pytest, hypothesis or
> Radicale versions may fail with bizzarre errors. This is especially a
> problem if packagers try to use their system-packages for installing
> test dependencies instead of just using a virtualenv and pip (I don't
> see a reason why that would be a problem TBH) See also #334 

Why not require specific versions in setup.py? You are already doing
that for some of your dependencies and I find it helpful.

> This seems to be the sanest approach to testing vdirsyncer as a
> packager: https://github.com/Homebrew/homebrew/pull/41208 i.e. just
> create an example config and run a simple sync. Even a simple
> `vdirsyncer --help` catches 80% of errors IMO.
> 
> ## Updating vdirsyncer
> 
> [The Hypothesis Packaging
> Guidelines](https://hypothesis.readthedocs.org/en/release/packaging.html)
> say that package maintainers who are unable to keep up with its
> release cycle are doing the community a disservice. While I think
> there are nicer ways to phrase that (especially towards people who are
> likewise doing a voluntary service), I am worried that vdirsyncer will
> find itself in a similarly problematic situation once [Debian starts
> to package
> it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=745027).
> Packages usually have been very quickly updated otherwise.

Agreed.

[...]

> ## Enforcing dependencies' minimum required versions
> 
> Vdirsyncer has a lower bound on the versions of some dependencies. As
> far as I understand those lower bounds have been completely ignored so
> far by downstream. Perhaps there's just no way to fix this situation.
> Again I'm more worried about Debian than any other distro here, but
> maybe it'll be fine.

Guix uses setup.py when building so we can't ignore your requirements
without patching setup.py. Also, it's trivial to have multiple package
versions in the distribution (also trivial for users to do this on their
own). This shouldn't be an issue for us.

I actually prefer upstream projects to clearly specify all dependencies
and their version ranges.

I'd like to work with you to keep providing a reliable and safe (i.e. up
to date) vdirsyncer package in Guix.

Respectfully awaiting your response, Leo

**Comment by lfam** [comment]
I don't know why GitHub is indenting part of my response incorrectly. Please don't skip it just because it's light grey :)

**Comment by mbaldessari** [comment]
I fundamentally agree with most of this. I think if someone is able to run the whole test suite and then file only vdirsyncer relevant bugs, let them, no big deal.

What would help me tremendously is a heads up when new dependencies are introduced in newer versions. That way I can start packaging them right away and won't be lagging behind as much. In terms of lower bound on dependency versions, I think that if they are all specified in either requirements.txt or setup.py, downstream will eventually pick them up.

Thanks for starting this discussion, much appreciated ;)

**Comment by eolianoe** [comment]
Comparing to others projects vdirsyncer is "easy" to package as the changelog and the setup.py are well described and the dependencies are clearly identified.

Regarding to the test discussion:
* will it be possible to have the list of dependencies needed to run the tests with a lower bound, as provided in the setup.py ?
* maybe, two targets can be defined in the Makefile. One to have some full tests (as the actual make test) and a second one just to test some very basic functions of vdisrycner.

**Comment by DamienCassou** [comment]
I'm talking for the nix package. Thanks for asking our opinion. Here are my answers.

> Many packagers try to run the full testsuite on a vdirsyncer installation, in order to find bugs with their setup. But: 99% of all packaging-related bugs are caught by seeing if `vdirsyncer --help` works correctly.

as a maintainer we can't know that. If you really feel that running vdirsyncer --help is enough, please make that the default test suite or document it so that packagers run only that. But in my opinion, that would be a bit sad: if you have a good test suite, we should run it to make sure everything is fine.

In all cases, please define a 'packaging test suite' running all tests you believe are important to run by GNU/Linux distributions.

> * It uses [hypothesis](http://hypothesis.readthedocs.org/) to randomly generate example input data. Automatic test generation is a really powerful tool for finding encoding-related bugs, but it also means that tests may randomly fail.

tests should be deterministic in my opinion. If I were you I would fix the tests. If you insist on having non deterministic tests, please remove them from the packaging test suite.

> * It launches CardDAV/CalDAV servers for integration testing. In other words, it's not just for finding bugs in vdirsyncer, but also elsewhere. Why would vdirsyncer-packagers care if a server introduced a bug that makes the tests fail?

I think it's important to know that there is a bug, wherever it is. Please remove these tests from the packaging test suite if you believe otherwise :-).

> * It requires very recent dependencies.

that is not a problem for nix. The only important thing is that these dependencies are clearly documented, preferably in the setup.py because this is where nix is looking for.

> Older pytest, hypothesis or Radicale versions may fail with bizzarre errors. This is especially a problem if packagers try to use their system-packages for installing test dependencies instead of just using a virtualenv and pip (I don't see a reason why that would be a problem TBH) See also #334

this is not a problem for nix if the dependency and its lower bound version are specified in setup.py.

> This seems to be the sanest approach to testing vdirsyncer as a packager: https://github.com/Homebrew/homebrew/pull/41208 i.e. just create an example config and run a simple sync. Even a simple `vdirsyncer --help` catches 80% of errors IMO.

could you please put that in your packaging test suite so we don't have to duplicate the exact same unit test in all distributions?

> [The Hypothesis Packaging Guidelines](https://hypothesis.readthedocs.org/en/release/packaging.html) say that package maintainers who are unable to keep up with its release cycle are doing the community a disservice. While I think there are nicer ways to phrase that (especially towards people who are likewise doing a voluntary service), I am worried that vdirsyncer will find itself in a similarly problematic situation once [Debian starts to package it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=745027). Packages usually have been very quickly updated otherwise.

not a problem with nix, we can update very fast. We just need to be notified of the change. We can be even faster if setup.py contains a list of all dependencies with minimum and maximum required versions.

> Apart from that we should probably provide an RSS feed for new releases. GitHub has a tags RSS feed that is mentioned [in the docs](https://vdirsyncer.readthedocs.org/en/latest/changelog.html), I guess we could just use that.

indeed. Please tell me when you have one and I will subscribe and update fast.

> Vdirsyncer has a lower bound on the versions of some dependencies. As far as I understand those lower bounds have been completely ignored so far by downstream.

nix pays attention to what is in setup.py. We can't ignore what is in setup.py except by patching it and we try not to.

**Comment by untitaker** [comment]
On Tue, Mar 01, 2016 at 02:28:42PM -0800, Hugo Osvaldo Barrera wrote:
> Regarding Debian, if we find that an older verision keeps getting packaged,
> and this results in lots of users reporting invalid bugs, maybe it would be
> best to actually *recommend* in the documentation that users install via
> pip+virtualenv.

We'll see. I'd like to avoid this situation preemptively rather than having to
write this recommendation into the docs.

**Comment by untitaker** [comment]
On Tue, Mar 01, 2016 at 08:28:25PM -0800, lfam wrote:
> You've given some reasons why packagers might have difficulty running
> your test suite, but I'm still not sure why you'd want us to stop. Is it
> creating extra work for you, or slowing down development in some way?
> It's very common for software projects to distribute their source code
> with a test suite.

Not at all, but I consider it a waste of *your* time trying to package all the
dependencies vdirsyncer's testsuite has. Especially obscure things like
pytest-subtesthack, a package that is virtually used by nobody except
vdirsyncer. In Nix and Guix new packages may come with less overhead than
elsewhere, but overhead nonetheless.

> In my experience, '--help' usually exercises a very small part of a code
> base. Is that not the case for vdirsyncer?

It does exercise only a small part, but (without having actual experience) I
suspect that part (setuptools entry points, importing etc) is the most common
compatibility problem when creating packages. I consider that sufficient,
perhaps you don't.

> Running a (good and up-to-date) test suite does help catch problems with
> our packaging, especially due to upstream portability bugs and version
> mis-matches, but it also helps upstream by finding more serious bugs.

The examples you cite are from software written in low-level languages.
Vdirsyncer, on the other hand, doesn't contain any CPU-architecture specific
code.

Its dependencies may, but e.g. lxml comes with its own testsuite, which is
covering more edgecases in lxml than our testsuite. I find it unlikely that
it'd be worth it to run vdirsyncer's testsuite to find bugs in lxml.

> Can you clarify what would be bad about these tests failing when we
> build these packages? Don't you want us to help you find those bugs?

Not at all.

> If you mean that the tests are non-deterministic, I suggest that it's
> not optimal to introduce non-determinism in the build process of your
> software (I know you think the tests aren't part of that process, but
> obviously I disagree). If you want to fuzz vdirsyncer, perhaps that
> belongs in a separate body of tests. Or, you could add a note to INSTALL
> instructing packagers to skip these tests.
> 
> Debian is already building >80% of their packages reproducibly [0], and
> in Guix we are creating infrastructure to start treating
> non-reproducible builds as a bug. I don't see this trend turning around;
> it's too important for reliability and security.
> 
> What do you think? Have I misunderstood your concern about this point?
> 
> [0] https://tests.reproducible-builds.org/reproducible.html

Hypothesis is actually much more deterministic than I implied (and really
doesn't take as much resources as fuzzing would do). In practice, I've never
seen CI randomly fail. I don't really see Hypothesis (or quickcheck in Haskell)
as a new kind of test anyway. It rather changes the way I would normally write
unit-tests.

For now I'd just see how it goes.

> 
> > * It launches CardDAV/CalDAV servers for integration testing. In other
> > words, it's not just for finding bugs in vdirsyncer, but also
> > elsewhere. Why would vdirsyncer-packagers care if a server introduced
> > a bug that makes the tests fail?
> 
> I see your point regarding this specific test. However, I don't think it
> extends to the rest of the test suite.
> 
> It's not very different from bugs being introduced in another one of
> vdirsyncer's build- or run-time dependencies. I want to know if an
> update to e.g. 'click' is going to break our vdirsyncer package, or
> expose a bug in vdirsyncer, etc. Nevertheless, as above, it could go in
> a separate test suite or come with a caveat in INSTALL.
> 
> What do you think?

The default `make test` does disable remoteStorage and DAV tests. But I've seen
people override those switches, and I don't think that's a good idea.

> 
> > * It requires very recent dependencies. Older pytest, hypothesis or
> > Radicale versions may fail with bizzarre errors. This is especially a
> > problem if packagers try to use their system-packages for installing
> > test dependencies instead of just using a virtualenv and pip (I don't
> > see a reason why that would be a problem TBH) See also #334 
> 
> Why not require specific versions in setup.py? You are already doing
> that for some of your dependencies and I find it helpful.

The dependencies change based on which DAV server or remoteStorage server you
test. I could specify the "base dependencies" with tests_require in setup.py,
but if you want to test a DAV server, you still need to run make install-test
(which executes a shell script that will invoke pip). Is this acceptable?

**Comment by untitaker** [comment]
On Wed, Mar 02, 2016 at 01:44:14AM -0800, eolianoe wrote:
> Comparing to others projects vdirsyncer is "easy" to package as the changelog and the setup.py are well described and the dependencies are clearly identified.
> 
> Regarding to the test discussion:
> * will it be possible to have the list of dependencies needed to run the tests with a lower bound, as provided in the setup.py ?

See the response to @lfam. To add to that response, the list of dependencies
depends on which servers you want to test, and currently `make test` figures
this out "on-the-go". So writing a definitive list of dependencies is only
possible for the "core" testsuite (which is what is run when you do `make
install-test test`).

> * maybe, two targets can be defined in the Makefile. One to have some full
> tests (as the actual make test) and a second one just to test some very basic
> functions of vdisrycner.

`make test` (without overriding DAV_SERVER and REMOTESTORAGE_SERVER) is
basically that. If you think that target is too slow, perhaps we need a subset
of that one.

**Comment by untitaker** [comment]
On Wed, Mar 02, 2016 at 12:16:45AM -0800, Michele Baldessari wrote:
> What would help me tremendously is a heads up when new dependencies are
> introduced in newer versions. That way I can start packaging them right away
> and won't be lagging behind as much. In terms of lower bound on dependency
> versions, I think that if they are all specified in either requirements.txt
> or setup.py, downstream will eventually pick them up.

This is a good point, and something I noticed while trying to write the CC list
for this issue's OP. I thought about using GitHub teams for that (since you can
CC teams as a whole), but I think some other mechanism is needed.

**Comment by untitaker** [comment]
On Wed, Mar 02, 2016 at 02:05:47AM -0800, Damien Cassou wrote:
> I'm talking for the nix package. Thanks for asking our opinion. Here are my
> answers.
> 
> > Many packagers try to run the full testsuite on a vdirsyncer installation,
> > in order to find bugs with their setup. But: 99% of all packaging-related
> > bugs are caught by seeing if `vdirsyncer --help` works correctly.
> 
> as a maintainer we can't know that. If you really feel that running
> vdirsyncer --help is enough, please make that the default test suite or
> document it so that packagers run only that. But in my opinion, that would be
> a bit sad: if you have a good test suite, we should run it to make sure
> everything is fine.
> 
> In all cases, please define a 'packaging test suite' running all tests you
> believe are important to run by GNU/Linux distributions.
> 
> > * It uses [hypothesis](http://hypothesis.readthedocs.org/) to randomly
> > generate example input data. Automatic test generation is a really powerful
> > tool for finding encoding-related bugs, but it also means that tests may
> > randomly fail.
> 
> tests should be deterministic in my opinion. If I were you I would fix the
> tests. If you insist on having non deterministic tests, please remove them
> from the packaging test suite.

Is that a point you insist on? In practice, Hypothesis' tests are not that
unpredictable. See also my previous replies -- I'd be tempted to just keep them
in and see how it goes.

> > * It launches CardDAV/CalDAV servers for integration testing. In other
> > words, it's not just for finding bugs in vdirsyncer, but also elsewhere.
> > Why would vdirsyncer-packagers care if a server introduced a bug that makes
> > the tests fail?
> 
> I think it's important to know that there is a bug, wherever it is. Please
> remove these tests from the packaging test suite if you believe otherwise
> :-).
> 
> > * It requires very recent dependencies.
> 
> that is not a problem for nix. The only important thing is that these
> dependencies are clearly documented, preferably in the setup.py because this
> is where nix is looking for.

How does this work? Are you looking at `tests_require`? Is the setup.py
evaluated like it would be when running `python setup.py`, or with some sort of
custom parser?

I am looking at
https://github.com/NixOS/nixpkgs/blob/40ca841c5c0b264a0d1a1ffc2f912e17e30f9caa/pkgs/tools/misc/vdirsyncer/default.nix#L13
and rather see a hardcoded list of dependencies.

> > Older pytest, hypothesis or Radicale versions may fail with bizzarre
> > errors. This is especially a problem if packagers try to use their
> > system-packages for installing test dependencies instead of just using a
> > virtualenv and pip (I don't see a reason why that would be a problem TBH)
> > See also #334
> 
> this is not a problem for nix if the dependency and its lower bound version
> are specified in setup.py.
> 
> > This seems to be the sanest approach to testing vdirsyncer as a packager:
> > https://github.com/Homebrew/homebrew/pull/41208 i.e. just create an example
> > config and run a simple sync. Even a simple `vdirsyncer --help` catches 80%
> > of errors IMO.
> 
> could you please put that in your packaging test suite so we don't have to
> duplicate the exact same unit test in all distributions?

Now I'm not really sure I should recommend that, seems like I have to
backpaddle on that one (see reply to @lfam).

**Comment by DamienCassou** [comment]
> I could specify the "base dependencies" with tests_require in setup.py, but if you want to test a DAV server, you still need to run make install-test (which executes a shell script that will invoke pip). Is this acceptable?

what is acceptable for nix, and probably guix too, is that all missing or bad dependencies are detected during the build phase. In nix, if the build phase tries to use pip, the build will stop with an error message such as "pip tried to download foobar version 1.6 but automatic downloading is not possible in nix to ensure reproducibility". At this point, the maintainer will manually add a dependency to the build process and try again. The build phase should detect that the dependency is present and not try to run pip again.

>>  tests should be deterministic in my opinion. If I were you I would fix the tests. If you insist on having non deterministic tests, please remove them from the packaging test suite.
> Is that a point you insist on?

very much yes. If building of vdirsyncer fails on nix' build farm, the build of the whole system stops and reports an error. The maintainer then has to spend time trying to understand what was the problem. The whole nixpkgs/nixos distribution will refuse to build until the maintainer fixes the build. Tests failing randomly are not acceptable for us.

>>> * It requires very recent dependencies.
>> that is not a problem for nix. The only important thing is that these dependencies are clearly documented, preferably in the setup.py because this is where nix is looking for.
>How does this work? Are you looking at `tests_require`? Is the setup.py evaluated like it would be when running `python setup.py`, or with some sort of custom parser?
> I am looking at https://github.com/NixOS/nixpkgs/blob/40ca841c5c0b264a0d1a1ffc2f912e17e30f9caa/pkgs/tools/misc/vdirsyncer/default.nix#L13 and rather see a hardcoded list of dependencies.

indeed. The process is what I explained a few lines above. Let me rephrase it:

1. the maintainer manually starts the build
2. the build fails because of a missing dependency (e.g., detected by running `setup.py` and catching the network accesses)
3. the maintainer manually add the missing dependency
4. go back to step 1 until the build succeeds

/cc @vcunat @domenkozar: is what I'm saying correct?

**Comment by 0-wiz-0** [comment]
I usually update the pkgsrc version of vdirsyncer within a week of the release.
When I notice it, I denote the lower bounds for dependencies, see the
[pkgsrc Makefile](http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/time/py-vdirsyncer/Makefile?rev=1.30&content-type=text/x-cvsweb-markup&only_with_tag=MAIN).

I usually do not run tests except for `--version` -- I'll notice if it's broken anyway at most a couple of days later when it goes into production on my systems.
Not sure if, but I hope this helps...

**Comment by untitaker** [comment]
On Wed, Mar 02, 2016 at 07:55:15AM -0800, Damien Cassou wrote:
> what is acceptable for nix, and probably guix too, is that all missing or bad
> dependencies are detected during the build phase. In nix, if the build phase
> tries to use pip, the build will stop with an error message such as "pip
> tried to download foobar version 1.6 but automatic downloading is not
> possible in nix to ensure reproducibility". At this point, the maintainer
> will manually add a dependency to the build process and try again. The build
> phase should detect that the dependency is present and not try to run pip
> again.

I think this is already the case. `make install-test` will install everything
required, `make test` runs the tests.

> >>  tests should be deterministic in my opinion. If I were you I would fix
> >>  the tests. If you insist on having non deterministic tests, please remove
> >>  them from the packaging test suite.
> > Is that a point you insist on?
> 
> very much yes. If building of vdirsyncer fails on nix' build farm, the build
> of the whole system stops and reports an error. The maintainer then has to
> spend time trying to understand what was the problem. The whole nixpkgs/nixos
> distribution will refuse to build until the maintainer fixes the build. Tests
> failing randomly are not acceptable for us.

I'll see what I can do.

> 1. the maintainer manually starts the build
> 2. the build fails because of a missing dependency (e.g., detected by running `setup.py` and catching the network accesses)
> 3. the maintainer manually add the missing dependency
> 4. go back to step 1 until the build succeeds

This is not at all taking lower-bounds on version numbers into account, if I
read the nix file correctly. An older version of dependency X than required
might get installed that way as well.

**Comment by DamienCassou** [comment]
>> 1. the maintainer manually starts the build
>> 2. the build fails because of a missing dependency (e.g., detected by running `setup.py` and catching the network
>> accesses)
>> 3. the maintainer manually add the missing dependency
>> 4. go back to step 1 until the build succeeds

>This is not at all taking lower-bounds on version numbers into account, if I read the nix file correctly. An older version of dependency X than required might get installed that way as well.

in step 2, the build should fail if the dependency version is not the one you expect. I should have written: "the build fails because of a missing or incorrect dependency"

**Comment by untitaker** [comment]
Hmm, crashing may not always be the case. I've seen versions fail in very bizzarre ways (hanging up processes and similar).

Anyway, I don't think it's that important. I'll post here if I have a draft of the guidelines. Thanks for the feedback!

**Comment by vcunat** [comment]
@DamienCassou: yes, the process may often look like that, but it depends... for example when I (personally) see that upstream documents what the package depends on, I start with that list.

**Comment by lfam** [comment]
On Wed, Mar 02, 2016 at 05:54:24AM -0800, Markus Unterwaditzer wrote:
> Not at all, but I consider it a waste of *your* time trying to package all the
> dependencies vdirsyncer's testsuite has. Especially obscure things like
> pytest-subtesthack, a package that is virtually used by nobody except
> vdirsyncer. In Nix and Guix new packages may come with less overhead than
> elsewhere, but overhead nonetheless.

I *did* grumble a bit when pytest-subtesthack was introduced. However,
given my reasons for wanting to run tests, I packaged it anyways :)

> It does exercise only a small part, but (without having actual experience) I
> suspect that part (setuptools entry points, importing etc) is the most common
> compatibility problem when creating packages. I consider that sufficient,
> perhaps you don't.
> 
> > Running a (good and up-to-date) test suite does help catch problems with
> > our packaging, especially due to upstream portability bugs and version
> > mis-matches, but it also helps upstream by finding more serious bugs.
> 
> The examples you cite are from software written in low-level languages.
> Vdirsyncer, on the other hand, doesn't contain any CPU-architecture specific
> code.

Perhaps those weren't the best examples to persuade you with. My point
was that it's not uncommon for software to be distributed as stable, but
still fail in strange ways when used on the veritable *myriad* of
environments found across the GNU / Linux world, let alone the other
platforms that Python supports. Test suites help catch that before users
do.

> Hypothesis is actually much more deterministic than I implied (and really
> doesn't take as much resources as fuzzing would do). In practice, I've never
> seen CI randomly fail. I don't really see Hypothesis (or quickcheck in Haskell)
> as a new kind of test anyway. It rather changes the way I would normally write
> unit-tests.
> 
> For now I'd just see how it goes.

Okay.

> The default `make test` does disable remoteStorage and DAV tests. But I've seen
> people override those switches, and I don't think that's a good idea.

I'm happy to skip the DAV test. I remember that I started running it
because the entire test suite would fail without it, breaking the build.
Has it been made optional?

> The dependencies change based on which DAV server or remoteStorage server you
> test. I could specify the "base dependencies" with tests_require in setup.py,
> but if you want to test a DAV server, you still need to run make install-test
> (which executes a shell script that will invoke pip). Is this acceptable?

It wouldn't work if pip wanted network access. Our build environments
are set up declaratively beforehand, and network access is not allowed
in order to enforce determinism.  Otherwise, it's just a matter of
providing those dependencies in the package definition.

**Comment by untitaker** [comment]
See #354. I think those address all points, except for the non-deterministic tests (unsure what to do. Another feature switch?)

**Comment by hobarrera** [comment]
Can we increase the amount of samples `hypothesis` tries, to a point where we can be "sure enough" that it won't fail after multiple runs? It's not like it's extremely random; it merely constructs objects based on certain rules, so a large enough amount of runs, we can can have enough certainty that it won't fail randomly any more.

The other options is to somehow exclude these tests (much like we have the option to skip webdav/remoteStorage tests, which is, IMHO, what should be recommended for packagers).

**Comment by untitaker** [comment]
Such settings are already in place IMO. I'll have to think about this a bit more.

**Comment by untitaker** [comment]
A question: Do you consider python-setuptools to be a dependency that should be explicitly listed, or is implicit okay, since it is part of pip, and pip is part of newer Python versions?

**Comment by mathstuf** [comment]
Just one note from me:

> Apart from that we should probably provide an RSS feed for new releases. GitHub has a tags RSS feed that is mentioned in the docs, I guess we could just use that.

Fedora has been using [anyita](https://release-monitoring.org/project/5624/) to track vdirsyncer updates. Other distributions are free to also track things there.

**Comment by hobarrera** [comment]
Actually, disabling hypothesis isn't the only way to make tests deterministic: we can also try to provide the same seed to hypothesis, so that is generates the same values.

Using the version number, or commit hash to seed hypothesis pre-test would make tests deterministic, while not reduce coverage.

**Comment by hobarrera** [comment]
I did some quick research, and indeed, this is doable via hypothesis [profiles](https://hypothesis.readthedocs.org/en/master/settings.html#settings-profiles) and using the `derandomize` setting.

I'd recommend having an environment flag like `DETERMINISTIC=true` which would trigger this profile. Pros of this choice:

 * Let us keep using hypothesis, and potentially find new edge cases.
 * Allow downstream to have their deterministic builds.
 * No need to change anything big (like how we run tests).

**Comment by untitaker** [comment]
@hobarrera Oh wow, I wasn't aware of that. #356 looks great, too.
